crates/fuel-streams-core/src/nats/stream.rs

```diff
@@ -0,0 +1,396 @@
+use std::fmt::Debug;
+
+use async_nats::jetstream::{
+    consumer::AckPolicy,
+    context::Publish,
+    kv,
+    stream,
+};
+use async_trait::async_trait;
+use fuel_data_parser::{
+    DataParser,
+    DataParserDeserializable,
+    DataParserSerializable,
+    NatsFormattedMessage,
+};
+use fuel_streams_macros::subject::IntoSubject;
+use tokio::sync::OnceCell;
+
+use super::{NatsError, NatsNamespace};
+use crate::{nats::types::*, prelude::NatsClient};
+
+// ------------------------------------------------------------------------
+// Traits
+// ------------------------------------------------------------------------
+
+#[async_trait]
+pub trait StreamEncoder:
+    Debug
+    + Clone
+    + Send
+    + Sync
+    + DataParserSerializable
+    + DataParserDeserializable
+    + 'static
+{
+    async fn stream_encode(
+        &self,
+        subject: &str,
+    ) -> Result<bytes::Bytes, NatsError> {
+        let data_parser = DataParser::default();
+        let encoded = data_parser.to_nats_payload(subject, self).await?;
+        Ok(encoded.into())
+    }
+
+    async fn stream_decode(
+        encoded: Vec<u8>,
+    ) -> Result<NatsFormattedMessage<Self>, NatsError> {
+        let data_parser = DataParser::default();
+        Ok(data_parser.from_nats_message(encoded).await?)
+    }
+}
+
+#[async_trait]
+pub trait Streamable:
+    Debug + Clone + Send + Sync + Sized + StreamEncoder
```

Comment:
These traits are already added above for StreamEncoder. I guess you dont need to repeat them here again, apart from the Sized!

Reply:
✅